srcts/src/shiny/bind.ts

```diff
@@ -75,33 +76,52 @@ const bindingsRegistry = (() => {
    * accessibility and other reasons. However, in practice our bindings still
    * work as long as inputs the IDs within a binding type don't overlap.
    *
-   * @returns ShinyClientError if current ID bindings are invalid, otherwise
-   * returns an ok status.
+   * @returns ShinyClientMessageEvent if current ID bindings are invalid,
+   * otherwise returns an ok status.
    */
   function checkValidity():
-    | { status: "error"; error: ShinyClientError }
+    | {
+        status: "error";
+        error: ShinyClientError;
+      }
+    | {
+        status: "warning";
+        event: ShinyClientMessageEvent;
+      }
     | { status: "ok" } {
     type BindingCounts = { [T in BindingTypes]: number };
     const duplicateIds = new Map<string, BindingCounts>();
+    let status = "ok";
+    const problems: Set<string> = new Set();
 
     // count duplicate IDs of each binding type
     bindings.forEach((idTypes, id) => {
       const counts: { [T in BindingTypes]: number } = { input: 0, output: 0 };
 
       idTypes.forEach((type) => (counts[type] += 1));
 
-      // If there's a single duplication of ids across both binding types, then
-      // when we're not in devmode, we allow this to pass because a good amount of
-      // existing applications use this pattern even though its invalid. Eventually
-      // this behavior should be removed.
-      if (counts.input === 1 && counts.output === 1 && !Shiny.inDevMode()) {
+      if (counts.input + counts.output < 2) {
         return;
       }
-
-      // If we have duplicated IDs, then add them to the set of duplicated IDs
-      // to be reported to the user.
-      if (counts.input + counts.output > 1) {
-        duplicateIds.set(id, counts);
+      // We have duplicated IDs, add them to the set of duplicated IDs to be
+      // reported to the user.
+      duplicateIds.set(id, counts);
+
+      // Duplicated IDs are now always a warning. Before the ShinyClient console
+      // was added duplicate output IDs were errors in "production" mode. After
+      // the Shiny Client console was introduced, duplicate IDs were no longer
+      // production errors but *would* break apps in dev mode. Now, in v1.10+,
+      // duplicate IDs are always warnings in all modes for consistency.
+      status = "warning";
+
+      if (counts.input > 1) {
+        problems.add("input");
+      }
+      if (counts.output > 1) {
+        problems.add("output");
+      }
+      if (counts.input >= 1 && counts.output >= 1) {
+        problems.add("shared");
       }
     });
 
```

Comment:
The line below this one should probably be

```
if (duplicateIds.size === 0) return { status: status };
```

Reply:
Another direction would be to get rid of the `status` variable altogether and use the length of `problems` to inform the early return / status